File: examples/nodejs/example01_load_version_logs.js
const {
  DXB_BACKEND_AUTO,
  DXB_FEATURE_DX12,
  DXB_LOG_LEVEL_NAMES,
  DXBridgeLibrary,
  getStringArg,
  parseArgs,
  parseBackend,
  parseVersion
} = require('./dxbridge');

function printUsage() {
  console.log('Usage: node examples/nodejs/example01_load_version_logs.js [--dll PATH] [--backend auto|dx11|dx12]');
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  if (args.values.help === true) {
    printUsage();
    return;
  }

  const backend = parseBackend(getStringArg(args, 'backend', 'auto'));
  const dllPath = getStringArg(args, 'dll');
  const dxbridge = new DXBridgeLibrary(dllPath || undefined);

  console.log(`Loaded: ${dxbridge.path}`);
  dxbridge.setLogCallback((level, message) => {
    const label = DXB_LOG_LEVEL_NAMES[level] || `LEVEL_${level}`;
    console.log(`[log:${label}] ${message}`);
  });

  try {
    const version = dxbridge.getVersion();
    console.log(`DXBridge version: ${parseVersion(version)} (0x${version.toString(16).toUpperCase().padStart(8, '0')})`);

    dxbridge.init(Number.isFinite(backend) ? backend : DXB_BACKEND_AUTO);
    console.log(`DX12 feature flag reported by active backend: ${dxbridge.supportsFeature(DXB_FEATURE_DX12)}`);
    console.log('Initialization and shutdown completed successfully.');
    dxbridge.shutdown();
  } finally {
    dxbridge.close();
  }
}

main();

File: examples/rust/src/bin/example01_load_version_logs.rs
#![cfg(windows)]

use dxbridge_rust_examples::cli::{AnyResult, CommandLine};
use dxbridge_rust_examples::dxbridge::{parse_backend, parse_version, DxBridge, DXB_FEATURE_DX12};

fn main() {
    if let Err(err) = run() {
        eprintln!("{}", err);
        std::process::exit(1);
    }
}

fn run() -> AnyResult<()> {
    let cli = CommandLine::from_env()?;
    if cli.help_requested {
        println!("Usage: example01_load_version_logs [--dll PATH] [--backend auto|dx11|dx12]");
        return Ok(());
    }

    let backend = parse_backend(&cli.get_string("--backend", "auto"))?;
    let bridge = DxBridge::load(cli.get_optional_string("--dll").as_deref())?;

    println!("Loaded: {}", bridge.path.display());
    bridge.set_log_callback(|level, message| {
        let label = match level {
            1 => "WARN",
            2 => "ERROR",
            _ => "INFO",
        };
        println!("[log:{}] {}", label, message);
    });

    let version = bridge.get_version();
    println!(
        "DXBridge version: {} (0x{:08X})",
        parse_version(version),
        version
    );
    bridge.init(backend)?;

    println!(
        "DX12 feature flag reported by active backend: {}",
        bridge.supports_feature(DXB_FEATURE_DX12)
    );
    bridge.shutdown();
    bridge.clear_log_callback();
    println!("Initialization and shutdown completed successfully.");
    Ok(())
}
